### PR TITLE
#167190528 Build Email Verification Service

### PR DESCRIPTION
**Pull request by chialuka**
#### What does this PR do?
- It adds the email verification function and checks to prevent unverified users after some time

#### Description of Task proposed in this pull request?
- Create an email module for sending any kind of email to users
- Create a template for email verification service
- Send verification emails when the user signs up
- Add route and controller to verify emails
- Prevent unverified users from access to the app after 24 hours of registration

#### How should this be manually tested (Quality Assurance)?
- Register a new account by making a `POST` request to the `/users/create/` endpoint. You will receive an email. Click the link in the email to verify your email.

#### What are the relevant pivotal tracker stories?
[[#167190528]](https://www.pivotaltracker.com/story/show/167190528)

#### Any background context you want to add (Operations Impact)?
- This PR adds email verification functionality. It allows the app to send and verify the emails of users after they have signed up. It also prevents users who haven't been verified from using the app.

#### What I have learned working on this feature: 
- I have learned how to create reusable email functions
- I have learned how to stub external libraries in testing

#### Screenshots: 
Screenshot 2019-08-07 at 8 57 42 PM
Screenshot 2019-08-07 at 8 58 01 PM

### CODE REVIEW

**Review by ayodejiAA — DISMISSED** [review]

**Review by ayodejiAA — DISMISSED** [review]

File: server/middlewares/verifyToken.js [hunk 1]
```diff
@@ -25,16 +25,29 @@ const verifyToken = async (request, response, next) => {
...
+ * @param {Object} request request object
+ * @param {Object} response express response object
+ * @param {Object} next express next function that calls the next middleware
+ * @returns {Void} it calls the next middleware
+ */
+const getSessionFromToken = async (request, response, next) => {
+  const { token } = response.locals;
+  const session = await findToken(token);
+  if (!session) {
+    return serverResponse(response, 440, {
```
Comment (dinobi) [inline server/middlewares/verifyToken.js]: What is `440`?

File: server/helpers/emailModule.js [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+sendGrid.setApiKey(process.env.SENDGRID_API_KEY);
+
+/**
+ * @name sendEmail
+ * @async
+ * @description function for sending emails to users
+ * @param {String} receiver email of recipient
+ * @param {String} subject subject of email to be sent
+ * @param {String} content text to be sent to user
+ * @returns {Boolean} true ffor successful email sending or error on failure
```
Comment (dinobi) [inline server/helpers/emailModule.js]: `ffor` typo

File: server/middlewares/verifyToken.js (hunk 1 above)
Comment (chialuka) [reply server/middlewares/verifyToken.js]: 440 is a Login timeout error

File: server/helpers/emailModule.js (hunk 2 above)
Comment (chialuka) [reply server/helpers/emailModule.js]: Ok. I'll fix it now

File: server/controllers/Users.js [hunk 3]
```diff
@@ -55,11 +56,36 @@ class Users {
...
       delete user.dataValues.password;
+      sendVerificationEmail({ ...user.dataValues, token });
       return serverResponse(res, 201, { user, token });
     } catch (error) {
       return serverError(res);
     }
   }
+
+  /**
+   * @name sendResetEmail
```
Comment (dinobi) [inline server/controllers/Users.js]: can you update this name

Comment (chialuka) [reply server/controllers/Users.js]: Right away

**Review by dinobi — APPROVED** [review]

**Review by ayodejiAA — APPROVED** [review]

**Review by henryade — APPROVED** [review]

**Pull request merged**